lib/Sema/TypeCheckProtocolInference.cpp

```diff
@@ -1440,17 +1469,6 @@ AssociatedTypeDecl *AssociatedTypeInference::inferAbstractTypeWitnesses(
 
       std::function<Type(Type)> substCurrentTypeWitnesses;
       substCurrentTypeWitnesses = [&](Type ty) -> Type {
-        if (auto *gp = ty->getAs<GenericTypeParamType>()) {
-          // FIXME: 'computeFixedTypeWitness' uses 'getReducedType',
-          // so if a generic parameter is canonical here, it's 'Self'.
-          if (gp->isCanonical() ||
-              isa<ProtocolDecl>(gp->getDecl()->getDeclContext()->getAsDecl())) {
-            return adoptee;
-          }
-
-          return ty;
-        }
-
```

Comment:
I forget the reason I moved it, but this part is hacky in general because it's relying on type sugar for semantic effect. Type witnesses should instead track if they're relative to the protocol's generic signature or the conformance's generic signature, instead of us trying to divine it from the Type.